rohmu/notifier/http.py

```diff
@@ -0,0 +1,98 @@
+"""Copyright (c) 2022 Aiven, Helsinki, Finland. https://aiven.io/"""
+from .interface import Notifier
+from dataclasses import dataclass
+from datetime import datetime
+from enum import Enum, unique
+from queue import SimpleQueue
+from typing import Any, Dict, Optional
+
+import requests
+import threading
+
+
+@unique
+class Operation(Enum):
+    UPLOAD = "UPLOAD"
+    DELETE = "DELETE"
+    DELETE_TREE = "DELETE_TREE"
+
+
+def _get_requests_session() -> requests.Session:
+    retry = requests.adapters.Retry(
+        total=3,
+        backoff_factor=0.5,
+        status_forcelist=[429, 500, 502, 503, 504],
+        method_whitelist={"POST"},
+    )
+    adapter = requests.adapters.HTTPAdapter(max_retries=retry)
+    request_session = requests.Session()
+    request_session.mount("http://", adapter)
+    return request_session
+
+
+class HTTPNotifier(Notifier):
+    def __init__(self, url: str) -> None:
+        self._session = _get_requests_session()
+        self._url = url
+
+    def object_created(self, key: str, size: Optional[int]) -> None:
+        self._session.post(
+            self._url,
+            {
+                "op": Operation.UPLOAD.value,
+                "key": key,
+                "size": size,
+                "last_modified": datetime.now(),
+            },
+        )
+
+    def object_deleted(self, key: str) -> None:
+        self._session.post(self._url, {"op": Operation.DELETE.value, "key": key})
+
+    def tree_deleted(self, key: str) -> None:
+        self._session.post(self._url, {"op": Operation.DELETE_TREE.value, "key": key})
+
+
+@dataclass
+class _HTTPNotifyJob:
+    url: str
+    data: Dict[str, Any]
+
+
+def _background_http_request(queue: "SimpleQueue[_HTTPNotifyJob]") -> None:
+    with _get_requests_session() as session:
+        while True:
+            notification = queue.get()
+            session.post(notification.url, notification.data)
```

Comment:
The notification is dropped if the `post` fails
Also,  it will also kill the thread?

Reply:
That the notification is dropped is ok, it's a notification we're doing best effort here (sessions is configured to retry 3 times), but we should probably just log it and you are right @sjamgade not kill the thread.

Reply:
dropping is fine, but something is also not difficult given that we are in a background thread
```suggestion
            try:
                 session.post(notification.url, notification.data)
            except SomeException: 
                     log_the_exception
                     time.sleep(5)
                     queue.put(notification)
```